add re-reviews to the MIR process

cpaelzer opened the pull request:
In the last sprint a thought that has come up is the general desire to take our mission of "quality in main" serious and that actually re-reviews of packages already in main would need to be part of that.

This starts to document what we could do to start the discussion as only about half the team was in the in-person session.

The rules are 60% subset of the iniital MIR rules and 10% slight modifications or small additions. It might make more sense to not create a second redundant template and instead somehow "mark" those that shall be "also done" or "only done" on re-reviews in the original template. Open for suggestions, as I said this is only meant to be a draft to get the ball rolling.

cpaelzer committed:
add rereviews to the MIR process

In the last sprint a thought that has come up is the general desire
to take our mission of "quality in main" serious and that actually
re-reviews of packages already in main would need to be part of that.

This starts to document what we could do to start the discussion
as only about half the team was in the in-person session.

The rules are 60% subset of the iniital MIR rules and 10% slight
modifications or small additions. It might make more sense to not create
a second redundant template and instead somehow "mark" those
that shall be "also done" or "only done" on re-reviews in the original
template. Open for suggestions, as I said this is only meant to be a
draft to get the ball rolling.

Signed-off-by: Christian Ehrhardt <christian.ehrhardt@canonical.com>

joalif reviewed (commented):
Overall I'm a +1 for the re-review process.
I believe we will  find a few packages missing tests.
I think we should discuss/decide how strict (or not) the rules shall be in that case, given
that the owning teams may not have the bandwidth to implement test suites.

dviererbe commented on README.md:
```diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+
+Right now this is **not active**, it is - so far - a documentation of the intent to do so helping all involved parties to better understand what this would mean.
+
+The MIR team and thereby in extension Ubuntu continuously evolves the MIR ruling, for example we add rules based on painful lessons learned to avoid the same to happen again. In that regard some asks for testing, quality and such might be best practise now, but have not been fulfilled by packages that are in main for a long time.
```
```suggestion
The MIR team and thereby in extension Ubuntu continuously evolves the MIR ruling. For example, we add rules based on painful lessons learned to avoid the same from happening again. In that regard the demand for testing, quality and so on might be best practice now, but have not been fulfilled by packages that are in main for a long time.
```

typo: `practise` -> `practice`
suggestion: slightly different punctuation/wording for better reading.

cpaelzer replied:
Integrated, thanks!

dviererbe commented on README.md:
```diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+3. As on the initial MIR review we might end up recommending a security (re-)review which goes into their queue at low prio
+
+## Special cases
+Special case - packges in main without an launchpad bug based MIR process.
```
```suggestion
Special case - packages in main without an launchpad bug based MIR process.
```

typo: `packges` -> `packages`

cpaelzer replied:
already had that fixed by now, thanks

dviererbe commented on README.md:
```diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+
+## Special cases
+Special case - packges in main without an launchpad bug based MIR process.
+Some older cases do not have anything for auditability, for those we'll reach out to the [owning team](http://reqorts.qa.ubuntu.com/reports/m-r-package-team-mapping.html) and ask to file an MIR as they should ask (and answer) themselfe the same questions we do on a modern MIR.
```
```suggestion
Some older cases do not have anything for auditability. For those packages we'll reach out to the [owning team](http://reqorts.qa.ubuntu.com/reports/m-r-package-team-mapping.html) and ask them to file an MIR as they should ask (and answer) themselfe the same questions we do on a modern MIR.
```

suggestion: slightly different punctuation/wording for better reading.

cpaelzer replied:
Integrated (slightly different than proposed)

dviererbe commented on README.md:
```diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+
+## Process
+
+We'd add (doing so is a TODO) a query that helps us to spot packages in main for the longest time without having a recent update on their MIR bugs - out of that pool we'd then assign in each weekly team meeting.
```
```suggestion
We would add (doing so is a TODO) a query that helps us to spot packages in main for the longest time without having a recent update on their MIR bugs. Out of that pool we'd then assign re-reviews of approved packages in each weekly MIR team meeting.
```

suggestion: write out spoken English and explicitly name the implicit subject for better readability

cpaelzer replied:
ack, done

dviererbe reviewed (approved):
praise: +1 for me, I really like the the Idea to re-evaluate the quality of packages in main (especially the old ones) so we can take action. Thanks for proposing this!

question: Is this a one-time effort, or should this evolve into a continual practice to regularly re-evaluate packages in main?

dviererbe reviewed (approved):
praise: +1 for me, I really like the the Idea to re-evaluate the quality of packages in main (especially the old ones) so we can take action. Thanks for proposing this!

question: Is this a one-time effort, or should this evolve into a continual practice to regularly re-evaluate packages in main?

slyon reviewed (commented):
I'm generally +1 as well.

But I'd like to integrate the re-review template with the existing review template, so they won't diverge. See my inline comments for suggestions.

didrocks reviewed (approved):
[once current typo/fixes are managed]

I really like it in general. I have slight concerns about the managements ack as some of us are not fully focused on distro work, but rather seen as best effort.
If the 1 per week covers either old MIR review OR new MIR, I think this is reasonable though.

cpaelzer commented:
@joalif 
> I think we should discuss/decide how strict (or not) the rules shall be in that case, given
> that the owning teams may not have the bandwidth to implement test suites.

Thanks, I think we want to push, but not too hard.
I've added some wording for that

cpaelzer commented on README.md:
```diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+## Progress
+
+We do not expect this to be quick, there are so many old packages in main that do not even have any trace how they got promoted.
+But if we never start it will never be better. So instead of holding back, we need to just start at some point.
```
ack, done

cpaelzer commented on README.md:
```diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+TODO-A: This does need a security re-review, so I'll assign ubuntu-security
+TODO-B: This does not need a security re-review
+
+Required TODOs:
```
They are, you quote the summary at the top to help people that only glance at it.
There is a whole [Security] section later which will decide what the decision here will be - that has explicit checks and rules.

I added a RULE: statement to explain.

cpaelzer commented:
> How are conditional ACKs handled? What leverage does the MIR team have, if conditions need to be met for a package to remain in main. e.g., what happens when ACK requirements have not been met a year after the review.

Thanks @eslerm - I have put some words to outline my expectation in regard to this into the new version.
Thanks for bringing up that aspect.

cpaelzer commented on README.md:
````diff
@@ -852,6 +852,150 @@ TODO-A: - TBD
+
+
+```
+TODO: Re-review for Package: TBDSRC
````
As discussed, we all liked this - it is part of v2 that I'll push soon.

cpaelzer committed:
Improve rereviews to the MIR process

Based on the great initial feedback this improves the PR by:
- Fixing various typos
- Adding improved wording in various places
- Defining a few aspects that were formerly unclear or missing
- Restructuring the rereview template to be part of the normal
  template avoiding redundancy

Signed-off-by: Christian Ehrhardt <christian.ehrhardt@canonical.com>

cpaelzer commented:
FYI: I pushed a new version which I hope addresses all of the feedback and discussion we had so far.

Maybe we can again get reviews (I marked all of you for review again) before next MIR meeting.
That way we can iterate this rather quickly into something we feel happy to fully support.

setharnold commented on README.md:
```diff
@@ -595,7 +667,10 @@ RULE: description and check if any of them is in main:
 RULE: indeed the same use case as the package this MIR is about.
-TODO: There is no other package in main providing the same functionality.
+RULE: On re-review finding duplication might lead to considerations to drop
+RULE: one of the two. They might not have been unifiably back then, but might
```
`unifiably` perhaps `interchangeable`? `equivalent`?

cpaelzer replied:
Completely rewritten, thanks!

setharnold commented on README.md:
```diff
@@ -834,12 +917,6 @@ OK:
-TODO: - no use of setuid
-TODO: - use of setuid, but ok because TBD (prefer systemd to set those
-TODO:   for services)
 TODO: - no important open bugs (crashers, etc) in Debian or Ubuntu
```
Reviewing bug history may show neglected packages, or packages that are too costly to maintain, etc. Maybe it needs its own bullet point, though?

cpaelzer replied:
We already have marked the checks of regular updates in upstream and in packaging as "to be looked at" for the re-review. But your thought is good. I do not think we need another check for it, but I added a hint right above the check what might be special on a re-review.

setharnold reviewed (approved):
Thanks for starting this, I really like the idea.

cpaelzer commented on README.md:
```diff
@@ -159,6 +159,16 @@ The reporter/reviewer is tasked to use the templates the following way:
+things only are important on a re-check.
+For that there is a section *[Re-Review]* which has entries only needed for
+that.
+Furthermore any `TODO:` entry that shall be re-checked ona re-review has
```
fixed, thanks!

cpaelzer commented on README.md:
```diff
@@ -546,20 +556,82 @@ By default statements are in the *OK* section, but issues that need to
+# Re-Reviewing an approved package
+
+The MIR team and thereby in extension Ubuntu continuously evolves the MIR
+ruling. For example, we add rules based on painful lessons learned to avoid
```
fixed, thanks!

cpaelzer commented on README.md:
```diff
@@ -546,20 +556,82 @@ By default statements are in the *OK* section, but issues that need to
+## Packages without a MIR bug
+
+This is a special case for things that are in main for quite some time.
+Those packages are in main without an launchpad bug based MIR process.
```
Fixed, thanks!

cpaelzer commented on README.md:
```diff
@@ -546,20 +556,82 @@ By default statements are in the *OK* section, but issues that need to
+Then we will reach out to the [owning team](http://reqorts.qa.ubuntu.com/reports/m-r-package-team-mapping.html)
+and ask them to look at what we found and also ask them to reply on the bug
+with a full new `MIR` template. That shall help to ensure the owning teams
+point of view is documented as they should ask (and answer) themselfe
```
Fixed, thanks!

cpaelzer committed:
Improve rereviews to the MIR process

Based on further feedback:
- add explanations where incomplete or misleading
- fix a few typos

Signed-off-by: Christian Ehrhardt <christian.ehrhardt@canonical.com>

slyon reviewed (approved):
lgtm.

Only thing would be adding the corresponding query to find the old packages in main with outdated or non-existing MIR bugs, which should be added to the meeting script.

dviererbe reviewed (approved):
praise: LGTM

seb128 commented:
Hey MIR team. I ended up here after reading yesterday's IRC meeting's log, unsure if there is a more proper place to have that discussion but to me it feels like we should have `owning teams` included before that new policy is put in place.

Sharing my personal initial thinking. While I understand the intend, it seems a ton of work that is going to be invested for really little benefit. Speaking for desktop we already struggle to find the resources to work on the package improvements needed to MIR the new packages we need promote, there is no way we will be able to take on extra tasks with our current resourcing.

I'm not going to start a discussion on the interest of debian/copyright, but even if we find resources to spend on improving our packages quality (which I think we should, at least for example on automated testing), updating debian/copyright is bring no value to our users and our product and I don't see us wanting to allocate resources to such tasks when we have and endless backlog of users visible issues/QA/workflow improvements to work on.

If you want to go ahead with that effort my preference would be to at least try to narrow a lot the checklist to focus on the things that will bring the most value back (build and autopkg tests, security review, ...)

I would be interested to see what other teams that will be on the 'receiving end' think about the proposal.

setharnold commented:
> updating debian/copyright is bring no value to our users and our product

We, as a company, need an SBOM solution soon to sell to some customers.

If debian/copyright files are consulted by the SBOM tools under consideration, we might need to get these into good shape very quickly.

If debian/copyright files aren't used by the SBOM tools under consideration, then any time spent on them feels 100% wasted. (At least, I've never seen any praise for them from my Debian circles, just complaints about how complex they are to prepare. Perhaps they do serve a useful purpose but it's well-hidden if so.)

eslerm commented on README.md:
```diff
@@ -546,20 +556,82 @@ By default statements are in the *OK* section, but issues that need to
+   1. But on the other hand this keeps effort of MIR team members somewhat plannable
+1. We'd run a trimmed down check which is a subset of our usual check (see below for details)
+   1. Anything we find for packages already in main becomes part of the owning teams backlog. The decision how hard we push on them is a case by case discussion depending on too many variables to properly pre-define.
+1. As on the initial MIR review we might end up recommending a security (re-)review which goes into their queue at low prio
```
please change "we might end up recommending a security (re-)review" to "we might assign a security (re-)review"

I also recommend removing "which goes into their queue at low prio". Priority of a Security MIR is set by the owning team relative to the owning teams other open security MIRs. We try to work on MIRs across all teams evenly, and having owning team set priority helps us know what is most relevant to them. It is exceedingly likely that an owning team will de-prioritize re-reviews though :)

eslerm commented on README.md:
```diff
@@ -546,20 +556,82 @@ By default statements are in the *OK* section, but issues that need to
+
+The proposed approach would be to:
+
+1. We will add (doing so is a TODO) a query that helps us to spot packages in main for the longest time without having a recent update on their MIR bugs.
```
Could re-review be avoided by comment bumping an MIR bug?

Re-assessing oldest MIRs first sounds fair, but may not be the most relevant. I don't have a better solution.

eslerm commented on README.md:
```diff
@@ -850,6 +931,12 @@ TODO-B: - translation present
 TODO-A: - TBD
 TODO-B: Problems: None
+
+[Re-Review*]
```
Should there be a requirement to see if upstream code is still maintained? e.g., that a git repo is not in archive mode

Perhaps this isn't `Re-Review` specific.

cpaelzer commented:
Nothing changed, but out of recent discussions it might make sense - even while unable to re-review all - to consider re-reviewing when a package undergoes major changes (all words that are hard to define well).

cpaelzer commented:
We have found our balance in re-reviews and landed it as optional. CLosing the PR - was a good discussion!
